Fixes #10526 - Support for Cisco POAP

fernandocarolo opened the pull request:
This adds support for Cisco PowerOn Auto Provisioning (POAP), a method to upgrade software images on Cisco Nexus switches running NX-OS. It has been tested with switches from the Nexus 6000 family, but should also work for other models, too.

A matching pull request for smart-proxy will be submitted shortly.

dLobatog commented:
[test] <- this triggers our Jenkins CI and runs tests with this PR, etc..

dLobatog commented on app/views/unattended/poap/POAP.erb:
```diff
@@ -0,0 +1,9 @@
+<%#
```
It looks like this would belong in docs rather than a default template. Assuming Cisco's are proprietary, do you mind maybe submitting a sample script of yours in [community-templates](https://github.com/theforeman/foreman/pull/2464/)?

fernandocarolo replied:
I am assuming Cisco scritps are proprietary since they are only distributed in their closed support site. I could not download them with a simple "guest" account not associated with a support contract.
I will write a new sample script in a couple of days, since I think that having a default template would simplify the use of this feature.

dLobatog commented on db/seeds.d/07-provisioning_templates.rb:
```diff
@@ -52,6 +53,8 @@
     { :name => "Junos default SLAX", :source => 'ztp/provision.erb', :template_kind => kinds[:provision], :operatingsystems => os_junos },
     { :name => "Junos default ZTP config", :source => 'ztp/ZTP.erb', :template_kind => kinds[:ZTP], :operatingsystems => os_junos },
     { :name => "Junos default finish", :source => 'ztp/finish.erb', :template_kind => kinds[:finish], :operatingsystems => os_junos },
+    { :name => "Junos default ZTP config", :source => 'ztp/ZTP.erb', :template_kind => kinds[:ZTP], :operatingsystems => os_junos },
```
Looks like you readded the ZTP config line by mistake here.

fernandocarolo replied:
You are right. Will remove it before resubmitting.

dLobatog commented on app/models/operatingsystems/nxos.rb:
```diff
@@ -0,0 +1,63 @@
+    "poap.cfg/"+host.mac.gsub(/:/,"").upcase
+  end
+
+  def kernel(arch)
```
Raise a `Foreman::Exception` here and in initrd perhaps if they are called, rather than returning none.

fernandocarolo replied:
Besides kernel and initrd, should I raise an exception at url_for_boot(file), too, since PXEFILES is empty?

fernandocarolo commented:
Ended up dropping the default POAP template after reading some Cisco scripts for different switch families. It looks like there no such thing as a "default" POAP script, there are lots of differences between models.

fernandocarolo committed:
Fixes #10526 - Support for Cisco POAP

dLobatog commented:
[test]

dLobatog commented:
@fernandocarolo Is the NXOS image corrupt? I don't see anything neither on github nor when I clone this branch. Also, could you dig up their trademark rules? It's possible that we're not allowed to use their pic and we have to use an alternate icon.

dLobatog commented on app/views/unattended/poap/disklayout.erb:
```diff
@@ -0,0 +1,5 @@
+kind: ptable
+name: NX-OS default fake
+%>
+# Not required for NX-OS POAP provisioning.
```
If the template is not required, just remove it? Default templates of any kind are normally added to [community-templates](https://github.com/theforeman/community-templates) then pulled here.

fernandocarolo replied:
I understand that a partition table is required for building a host, and it should be associated with the operating system family, even with the installer does not use it. There are currently two fake partition table templates, one for CoreOS and one for Junos, so I ended up using the same approach and included a dummy partition template for NX-OS.
Could you please confirm that adding a simple, dummy template to [community-templates](https://github.com/theforeman/community-templates) as a convenience to other users would be the right thing?

dLobatog replied:
@fernandocarolo That sounds fair enough, basically the same approach as ZTP. Please submit a PR to community-templates with that then :smile:

dLobatog commented:
@fernandocarolo Almost good to go in my opinion, although I definitely do not have the hardware to test this. 

Please just take the template and submit it to community-templates, and substitute the NXOS.png image by another image if possible. I'm not sure why but it seems like it's a 0kb file no matter how I look at it (downloaded it, tried to open it through Github, etc...)

Thanks again, almost there :cat: !

fernandocarolo commented:
@eLobato I think all your points should be addressed now:
- Fixed the NXOS.png file, since the previous one was corrupted due to a silly git mistake. Note that the logo is a simple placeholder, since I do not believe that NX-OS have a logo and I would not have permission to use it anyway.
- Removed the provisioning templates and will submit a pull request on [community-templates](https://github.com/theforeman/community-templates).
- I am raising a `Foreman::Exception` on `NXOS.url_for_boot(file)`, but if I do the same for `kernel` and `initrd` it breaks `Orchestration::TFTP.generate_pxe_template`. I ended up leaving dummy return values for those methods.

dLobatog commented:
[test]

dLobatog commented:
Merged as d9c57b3b6868e2e6ce206badd4d9f574303ef5ee, thanks @fernandocarolo!
